File: 0063-unique-paths-ii/0063-unique-paths-ii.go
# @param {Integer[][]} obstacle_grid
# @return {Integer}
def unique_paths_with_obstacles(obstacle_grid)
  m = obstacle_grid.size
  n = obstacle_grid.first.size
  row = [0] * n
  for i in (0...n)
    if obstacle_grid.first[i] == 1
      break
    else
      row[i] = 1
    end
  end
  for i in (1...m)
    for j in (0...n)
      row[j] = obstacle_grid[i][j] == 0 ? (j > 0 ? row[j - 1] : 0) + row[j] : 0
    end
  end
  row.last
end

File: 0063-unique-paths-ii/0063-unique-paths-ii.php
class Solution {

    /**
     * @param Integer[][] $obstacleGrid
     * @return Integer
     */
    function uniquePathsWithObstacles($obstacleGrid) {
        $row = count($obstacleGrid);
        $column = count($obstacleGrid[0]);
        for ($i = 0; $i < $row; $i++) {
            for ($j = 0; $j < $column; $j++) {
                if ($obstacleGrid[$i][$j] == 1) {
                    $obstacleGrid[$i][$j] = 0;
                } else if ($i == 0 && $j == 0) {
                    $obstacleGrid[$i][$j] = 1;
                } else {
                    $obstacleGrid[$i][$j] = $obstacleGrid[$i - 1][$j] + $obstacleGrid[$i][$j - 1];
                }
            }
        }
        return $obstacleGrid[$row - 1][$column - 1];
    }
}